internal/webhook/shared/helpers.go

```diff
@@ -0,0 +1,71 @@
+package shared
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// NewLogConstructor returns a log constructor function that adds the webhook name to the logger context for each admission request.
+func NewLogConstructor(name string) func(logr.Logger, *admission.Request) logr.Logger {
+	return func(_ logr.Logger, req *admission.Request) logr.Logger {
+		base := ctrl.Log
+		l := admission.DefaultLogConstructor(base, req)
+		return l.WithValues("webhook", name)
+	}
+}
+
+// CountObjects returns the number of objects of the given GroupVersionKind in the cluster using the provided client.
+// It lists all objects of the specified GVK and returns their count, or an error if the list operation fails.
+func CountObjects(ctx context.Context, cli client.Client, gvk schema.GroupVersionKind) (int, error) {
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(gvk)
+
+	if err := cli.List(ctx, list); err != nil {
+		return 0, err
+	}
+
+	return len(list.Items), nil
+}
+
+// DenyCountGtZero denies the admission request if there is at least one object of the given GroupVersionKind in the cluster.
+// It returns an admission.Denied response with the provided message if the count is greater than zero, otherwise it allows the request.
+// If there is an error counting objects, it returns an Errored response.
+func DenyCountGtZero(ctx context.Context, cli client.Client, gvk schema.GroupVersionKind, msg string) admission.Response {
+	count, err := CountObjects(ctx, cli, gvk)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	if count > 0 {
+		return admission.Denied(msg)
+	}
+
+	return admission.Allowed("")
+}
+
+// ValidateDupCreation denies creation if another instance of the same kind already exists (singleton enforcement).
+// It returns an Errored response if the kind does not match, otherwise it delegates to DenyCountGtZero.
+func ValidateDupCreation(ctx context.Context, cli client.Client, req *admission.Request, expectedKind string, log logr.Logger) admission.Response {
```

Comment:
I'd recommend to pass the logger as part of the context instead of an explicit argument